btc/cpfp.go

```diff
@@ -0,0 +1,476 @@
+			zap.Any("sequencesMap", sequencesMap),
+			zap.Int("fee", fee),
+			zap.Int("feeOverhead", feeOverhead),
+			zap.Int("feeRate", feeRate),
+			zap.Int("depth", depth),
+		)
+		return nil, ErrBuildCPFPDepthExceeded
+	}
+
+	var spendUTXOs UTXOs
+	var spendUTXOsMap map[btcutil.Address]UTXOs
+	var balanceOfScripts int64
+	var err error
+
+	// Get UTXOs for spend requests
+	err = withContextTimeout(c, 5*time.Second, func(ctx context.Context) error {
+		spendUTXOs, spendUTXOsMap, balanceOfScripts, err = getUTXOsForSpendRequest(ctx, w.indexer, spendRequests)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	spendUTXOsMap[w.address] = append(spendUTXOsMap[w.address], utxos...)
+	if sequencesMap == nil {
+		sequencesMap = generateSequenceMap(spendUTXOsMap, spendRequests)
+	}
+
+	// Check if there are no funds to spend for the given scripts
+	if balanceOfScripts == 0 && len(spendRequests) > 0 {
+		return nil, fmt.Errorf("scripts have no funds to spend")
+	}
+
+	// Temporary send requests for the transaction
+	tempSendRequests := append([]SendRequest{}, sendRequests...)
+
+	// If there are UTXOs and no spend/send requests, create a self-send request
+	if len(utxos) > 0 && len(tempSendRequests) == 0 && len(spendRequests) == 0 {
+		amount := int64(0)
+		for _, utxo := range utxos {
+			amount += utxo.Amount
+		}
+		tempSendRequests = append(sendRequests, SendRequest{
+			Amount: amount - int64(fee+feeOverhead),
+			To:     w.address,
+		})
+	}
+
+	// Build the transaction with the available UTXOs and requests
+	tx, signIdx, err := buildTransaction(append(spendUTXOs, utxos...), sacps, tempSendRequests, w.address, int64(fee+feeOverhead), sequencesMap)
+	if err != nil {
+		return nil, err
+	}
+
+	swSigs, trSigs := getNumberOfSigs(spendRequests)
+	bufferFee := 0
+	if depth > 0 {
+		// buffer fee accounts for
+		// 4 bytes for each sewgwit signature in the spend requests
+		// 4 bytes per each UTXO in the spend requests
+		// 1 byte for each taproot signature in the spend requests
+		// /2 is to convert the bytes to virtual size
+		bufferFee = ((4*(swSigs+len(utxos)) + trSigs) / 2) * feeRate
+	}
+
+	// Sign the spend inputs
+	err = withContextTimeout(c, 5*time.Second, func(ctx context.Context) error {
+		return signSpendTx(ctx, tx, signIdx, spendRequests, spendUTXOsMap, w.indexer, w.privateKey)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Sign the fee providing inputs, if any
+	err = signSendTx(tx, utxos, len(spendUTXOs), w.address, w.privateKey)
+	if err != nil {
+		return tx, err
+	}
+
+	// Calculate the true size of the transaction
+	txb := btcutil.NewTx(tx)
+	trueSize := mempool.GetTxVirtualSize(txb)
+
+	// Estimate the new fee
+	newFeeEstimate := (int(trueSize) * (feeRate)) + feeOverhead + bufferFee
+
+	// If the new fee estimate exceeds the current fee, rebuild the CPFP transaction
+	if newFeeEstimate > fee+feeOverhead {
+		buf := bytes.NewBuffer(make([]byte, 0, tx.SerializeSize()))
+		err = tx.Serialize(buf)
```

Comment:
Can you extract getting bytes from `wire.MsgTx` into a separate function and use it everwhere.